Fix several bugs related to scrolling

pyfisch opened the pull request:
* scrollLeft/scrollTop returned values of parent or even document root
   Only the scroll of the node itself is returned. Otherwise 0.0.
* Scrolling via script had set viewport.
   This resulted in other nodes appearing scrolled.
   Now scroll_offsets are updated with correct node id.

These bugs caused other odd behavior like both body and
document.documentElement being scrolled or the view for scrolled
elements jumping.

Also try scrolling this [example page](https://pyfisch.org/stuff/scrolltest.html) in servo with and without this change.

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #17342 (github issue number if applicable).

<!-- Either: -->
- [x] There are tests for these changes (partially)

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17385)
<!-- Reviewable:end -->

pyfisch committed:
Fix several bugs related to scrolling

* scrollLeft/scrollTop returned values of parent or even document root
   Only the scroll of the node itself is returned. Otherwise 0.0.
* Scrolling via script had set viewport.
   This resulted in other nodes appearing scrolled.
   Now scroll_offsets are updated with correct node id.

These bugs caused other odd behavior like both body and
document.documentElement being scrolled or the view for scrolled
elements jumping.

highfive commented:
Heads up! This PR modifies the following files:
 * @fitzgen: components/script/dom/element.rs, components/script/dom/window.rs
 * @kichjang: components/script/dom/element.rs, components/script/dom/window.rs

pyfisch commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 6ab0a71740532a2f4b4528bd0537639bd49ef2f3 with merge f4876f0dd93cac312aa8595ea183001589b64ea6...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/7544)

pyfisch commented:
The tests fail because I modified one web-platform-test to include more checks. Should web platform tests not be changed or do I need to update the manifest?

stshine commented:
r? @mrobinson

jdm commented:
./mach update-manifest

pyfisch commented:
Thanks. Ran `./mach update-manifest`.

pyfisch commented on tests/wpt/metadata/MANIFEST.json:
```diff
@@ -169897,6 +169897,66 @@
      {}
     ]
    ],
+   "_certs/064162.pem": [
```
Should these lines really be here? To me they look more like some left-over temporary files.

mrobinson commented on components/script/dom/window.rs:
```diff
@@ -1450,33 +1446,29 @@ impl Window {
             return;
         }
+
+        self.scroll_offsets.borrow_mut().insert(node.to_untrusted_node_address(),
```
Currently these scroll offsets come from WebRender. What is the issue here with waiting until WebRender updates the scroll offset map?

pyfisch replied:
Webrender does not update the scroll offset map if the change is caused by a `Msg::UpdateScrollStateFromScript` (per documentation). So the offset map needs to updated in script. The reason I think is that if a script does `elem.scrollTop = 42` the next line may be `console.log(elem.scrollTop)`. If there is some message-passing involved the log statement will print wrong information. (Another case where this is important is when a website sets scrollTop and scrollLeft at the same time as done in the tests).

mrobinson replied:
Hrm. The next time that WebRender sends new scroll offsets, the offsets really should reflect this change (otherwise that is a bug, I think). You are right though, that if we query the offsets immediately they won't reflect the new status. That's probably worth writing in a comment there.

pyfisch replied:
I don't think it is a bug since [it is described](https://doc.servo.org/script_layout_interface/message/enum.Msg.html#variant.UpdateScrollStateFromScript). If you scroll with mouse or keyboard the scroll state is updated by webrender. It also takes the changes from the script into account that happened before.

(Setting the offsets from webrender again in this case could be harmful. It may happen that webrender overwrites newer values.)

mrobinson replied:
Changes in scroll position need to be forwarded to WebRender to proper render the scrolled content. WebRender should then return the new offsets, but they won't be available immediately. I'm not sure the documentation touches on this.

pyfisch replied:
Yes this is how it is done for normal scrolling with mouse and keyboard. I  think the note on the enum member is sufficient but maybe you want to document it other places too or give more details.

mrobinson commented on tests/wpt/metadata/MANIFEST.json:
```diff
@@ -397231,6 +397291,54 @@
    "b960bef807da94c0146ed2f537eaa1e05ec9a0ab",
    "testharness"
   ],
+  "_certs/064162.pem": [
```
The update-manifest script isn't smart enough to ignore files that it shouldn't include, so you'll need to remove all of these extraneous lines manually. Sorry, I find this behavior really annoying too!

pyfisch replied:
Thanks for the explanation. I will do so.

pyfisch commented:
Added a comment and removed lines from manifest.

@mrobinson can you please r+?

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 5a87833359373d7c6be592b4efb0b96fe7ad7775 with merge dd1230f33443231864af970e48a7c6bf787d3c6e...

mrobinson reviewed (approved):
Thanks! This looks good.

mrobinson commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 5a87833 has been approved by `mrobinson`

<!-- @bors-servo r=mrobinson 5a87833359373d7c6be592b4efb0b96fe7ad7775 -->

bors-servo commented:
:hourglass: Testing commit 5a87833359373d7c6be592b4efb0b96fe7ad7775 with merge 2ff7ce783a9c144b80be492f3d0ad5468fc26edc...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4654)

jdm commented:
```
  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/scroll-to-fragid/003.html:
  │ FAIL [expected PASS] Fragment Navigation: Updating scroll position
  │   → assert_true: expected true got false
  │ 
  │ @http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/003.html:17:3
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  └ @http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/003.html:12:1

  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-percent-encoded.html:
  │ FAIL [expected PASS] Fragment Navigation: fragment id should be percent-decoded
  │   → assert_equals: expected 200 but got 0
  │ 
  │ steps<.handler@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-percent-encoded.html:17:9
  │ runNextStep/listener<@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-percent-encoded.html:40:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20

  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-anchor-name.html:
  │ FAIL [expected PASS] Fragment Navigation: scroll to anchor name is lower priority than equal id
  │   → assert_equals: expected 200 but got 0
  │ 
  │ steps<.handler@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-anchor-name.html:16:9
  │ runNextStep/listener<@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-anchor-name.html:34:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20

  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-id-top.html:
  │ FAIL [expected PASS] Fragment Navigation: TOP is a valid element id, which overrides navigating to top of the document
  │   → assert_equals: expected 200 but got 0
  │ 
  │ steps<.handler@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-id-top.html:14:7
  │ runNextStep/listener<@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-id-top.html:32:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20

  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-top.html:
  │ FAIL [expected PASS] Fragment Navigation: When fragid is TOP scroll to the top of the document
  │   → assert_not_equals: got disallowed value 0
  │ 
  │ steps<.handler@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-top.html:14:7
  │ runNextStep/listener<@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-top.html:41:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20
```

pyfisch commented:
```
▶ Unexpected subtest result in /html/browsers/browsing-the-web/scroll-to-fragid/003.html:
  │ FAIL [expected PASS] Fragment Navigation: Updating scroll position
  │   → assert_true: expected true got false
  │ 
  │ @http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/003.html:17:3
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  └ @http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/003.html:12:1
```
This is a standards mode document (`<!doctype html>`). It is scrolled to some fragments. Then `document.body.scrollTop` is requested. It should be nonzero if the document is scrolled. [The spec](https://drafts.csswg.org/cssom-view/#dom-element-scrolltop) states that this is only true in *quirks mode*, in standards mode *document.documentElement* is scrolled. (The test passes if one replaces `document.body` with `document.documentElement`). **Test is wrong.** (This passed before because servo scrolled both body and root element.)
```
  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-percent-encoded.html:
  │ FAIL [expected PASS] Fragment Navigation: fragment id should be percent-decoded
  │   → assert_equals: expected 200 but got 0
  │ 
  │ steps<.handler@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-percent-encoded.html:17:9
  │ runNextStep/listener<@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-percent-encoded.html:40:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20

  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-anchor-name.html:
  │ FAIL [expected PASS] Fragment Navigation: scroll to anchor name is lower priority than equal id
  │   → assert_equals: expected 200 but got 0
  │ 
  │ steps<.handler@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-anchor-name.html:16:9
  │ runNextStep/listener<@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-anchor-name.html:34:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20

  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-id-top.html:
  │ FAIL [expected PASS] Fragment Navigation: TOP is a valid element id, which overrides navigating to top of the document
  │   → assert_equals: expected 200 but got 0
  │ 
  │ steps<.handler@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-id-top.html:14:7
  │ runNextStep/listener<@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-id-top.html:32:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20

  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-top.html:
  │ FAIL [expected PASS] Fragment Navigation: When fragid is TOP scroll to the top of the document
  │   → assert_not_equals: got disallowed value 0
  │ 
  │ steps<.handler@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-top.html:14:7
  │ runNextStep/listener<@http://web-platform.test:8000/html/browsers/browsing-the-web/scroll-to-fragid/scroll-to-top.html:41:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1430:20
```
These failures are real bugs caused by me removing the line `self.update_viewport_for_scroll(x, y);`. I am still not sure how `perform_a_scroll` works on elements, maybe someone can explain it? But it is necessary to do the scrolls.

I have updated the pull request to disable the first test and added `update_viewport_for_scroll` at two places.

jdm commented:
Let's file an issue upstream about 003.html in that case.

pyfisch commented:
Filed an issue at w3c/web-platform-tests#6289.

pyfisch commented:
Is there anything still blocking this PR? The broken test is broken in Firefox too so it no servo bug. In Chrome it will work because of a [known and soon fixed incompatibility](https://bugs.chromium.org/p/chromium/issues/detail?id=157855).

mrobinson commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 284cb8a has been approved by `mrobinson`

<!-- @bors-servo r=mrobinson 284cb8aae8ed0c01c12aa866e8e7b15b07fcb816 -->

bors-servo commented:
:evergreen_tree: The tree is currently closed for pull requests below priority 9000, this pull request will be tested once the tree is reopened

bors-servo commented:
:hourglass: Testing commit 284cb8aae8ed0c01c12aa866e8e7b15b07fcb816 with merge 626c029623fea999fd7ea8650a6df8859ac87fcd...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7657), [arm32](http://build.servo.org/builders/arm32/builds/7876), [arm64](http://build.servo.org/builders/arm64/builds/7912), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8002), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4756), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4789), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7989), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/469), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/466), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4696), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4754), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/467), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/472), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3234)
Approved by: mrobinson
Pushing 626c029623fea999fd7ea8650a6df8859ac87fcd to master...
